src/main/kotlin/me/carleslc/kotlinextensions/collections/CollectionExtensions.kt

```diff
@@ -80,3 +97,21 @@ inline fun <T> Collection<T>.secondHalf(): List<T> = drop(half)
 
 inline fun <T> Collection<T>.split(index: Int): Pair<List<T>, List<T>> = take(index) to drop(index)
 inline fun <T> Collection<T>.split(): Pair<List<T>, List<T>> = split(half)
+
+fun uniqueRandoms(n: Int, range: LongRange): Set<Long> {
+
+    if (range.size < n) throw IllegalArgumentException("$n unique numbers not possible between $range, select a bigger range or reduce the number of unique numbers required.")
+
+    val uniqueRands = mutableSetOf<Long>()
+    val gap = range.size / n
+    var next = range.random()
+
+    for (i in 0 until n) {
+        next += nextLong(1, gap + 1)
```

Comment:
This is my proposal for an evenly-distributed unique random distribution:

```kotlin
fun uniqueRandoms(n: Int, range: LongRange = Long.MIN_VALUE .. Long.MAX_VALUE): Set<Long> {
    val elements = n.toBigInteger()

    if (range.size < elements) throw IllegalArgumentException("$n unique numbers not possible between $range, select a bigger range or reduce the number of unique numbers required.")

    val uniqueRands: Sequence<Long> = if (elements > range.size / BigInteger.valueOf(2)) {
        range.shuffled().asSequence()
    } else {
        generateSequence { range.random() }.distinct()
    }

    return uniqueRands
        .take(n)
        .toSet()
}
```

Two algorithms are in use here. One generates random values in the range and if a value is repeated then is discarded (sequence `distinct`) and a new value is retrieved. Collisions are more probable when more than 50% of range elements are going to be retrieved, so in that case, it's faster to just shuffle randomly the range and take the first n elements sequentially. If more than 50% of range elements are going to be taken anyway then the memory used to shuffle the entire range is not an issue.

Note that I'm using the `LongRange.size` method returning a `BigInteger` as I comment below to handle all Long ranges without overflow.

Playground: https://pl.kotl.in/RNL_B_HcB

Reply:
You did raise some very valid points. I do like the output of the correct algorithm more.

I wrote this algorithm to generate n number of random numbers in O(n) time. I timed the algorithm you sent me and it is taking twice as long as mine. 

It is up to you which versions out of the two should be kept, on the other hand, these little performances don' matter a lot of languages like Kotlin. Let me know which version out of the two you need and I will make the necessary changes.

Reply:
It makes sense that the second algorithm takes twice as long to run, as your algorithm is O(n) and my modification is around O(2n), but in fact, both are asymptotically equivalent complexities as O(c*n) ~ O(n), and we gain in correctness. So both algorithms are O(n), the difference is only in a constant factor which is ignored in time complexity analysis. Also in this case, as n is increasing, heap memory to store n elements is an issue sooner than the time performance for both algorithms.

I prefer correctness over performance, especially when comparing algorithms in the same asymptotical complexity. Small constant-factor performance improvements are not relevant enough if the output is not correct enough (in this case, a uniform random distribution*).
_*true randomness can never be achieved in a computer, random number generators as `range.random` or `nextLong` are always pseudo-random_, but it's better if our random algorithm does not induce even less randomness, so in the case range.random was truly random our algorithm should be truly random as well, which is the case for the second algorithm but not for the first because of its diminishing probabilities.

This library is not intended to be _very_ performant, but easy to use with the goal of improving programmer's productivity, with the functions being correct and performant enough for the majority of cases. If a function performance is very important for a specific project to the point of improving constant factors then it's better they have their custom function suited to their specific project needs.

Reply:
Explaining the analysis. As my proposal algorithm has two branches with two different algorithms, this is the complexity analysis:

**if n > range.size / 2 (retrieve more than 50% of range elements)
take n from `range.shuffled()`**

Space required is O(range.size + n) as all range elements are converted to a list when shuffling, and then it returns a set with the n first elements. In this case, that is O(n+n) = O(2n) when n is equal to range.size, which is asymptotically equivalent to O(n).
So, space is **O(n)**.

Time is O(range.size) for the range to a mutable list, O(range.size) for shuffle, and O(n) for take n. So with n = range.size is O(n+n+n) = O(3n) ~ O(n).
So, time is **O(n)**.

**if n <= range.size / 2
take n distinct from `range.random()`**

Space required is O(n) for the intermediate set in the stateful distinct() method, and O(n) to fill the returning set with n elements. O(n+n) = O(2n) ~ O(n). It could be improved to reuse the same set, but this way is more readable (don't have to reimplement the distinct method) and probably the improvement wouldn't be relevant.
So, space is **O(n)**.

Time depends on collisions for each element. If the element already exists, then another is generated. If this were the only algorithm then on average every element would take log(n) [20 for one million elements, 31 for n = Int.MAX_VALUE] attempts on average to get one distinct element by random and the algorithm would be O(n\*log(n)), but because we cut this to maximum 50% of collision probability (for more than 50% the first algorithm is used) then each element would need 2 attempts on average. For the first element only 1 attempt will succeed, but as we approach the 50% of elements already added the probability of a collision increases. With 50% the expected needed attempts are 1\*0,5+2\*0,5^2+3\*0,5^3+4\*0,5^4+5\*0,5^5... That sum converges to 2, so the expected time for taking n random distinct elements is O(2n) ~ O(n). Worst case scenario is always choosing an existing element and never returning, O(∞), but it is so unlikely is irrelevant. With as many range.random calls as all atoms in the universe (~10^82) that would be a chance of 273 consecutive collisions one time for just a single element (probability 1/2^273). And even in that case the time it takes to generate ~300 random elements is imperceptible. Can't even imagine the situation where it never ends (1/2^∞), i.e. almost 0 probability.
So, time is **O(n)**.

In both cases space and time is O(n), only with some small hidden constants as difference, so in summary the algorithm complexity is O(n).

It's not wonderful but is still O(n) and with a more uniform random distribution.